src/xpcc/driver/rtc/ds1302/ds1302.hpp

```diff
@@ -0,0 +1,153 @@
+// coding: utf-8
+/* Copyright (c) 2017, Sascha Schade (strongly-typed)
+ * All Rights Reserved.
+ *
+ * The file is part of the xpcc library and is released under the 3-clause BSD
+ * license. See the file `LICENSE` for the full license governing this code.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC_DS1302_HPP
+#define XPCC_DS1302_HPP
+
+namespace xpcc
+{
+
+struct ds1302
+{
+public:
+	struct xpcc_packed
+	Data
+	{
+		inline uint8_t
+		getSeconds() {
+			return doubleBcdToDecimal(data[0]);
+		}
+
+		inline uint8_t
+		getMinutes() {
+			return doubleBcdToDecimal(data[1]);
+		}
+
+		inline bool
+		is24hours() {
+			return not (data[2] & (1 << 7));
+		}
+
+		inline bool
+		isPm() {
+			return data[2] & (1 << 5);
+		}
+
+		inline bool
+		isAm() {
+			return not isPm();
+		}
+
+		inline uint8_t
+		getHours() {
+			if (is24hours()) {
+				return doubleBcdToDecimal(data[2]);
+			} else {
+				return doubleBcdToDecimal(data[2] & 0x1f);
+			}
+		}
+
+		inline uint8_t
+		getDate() {
+			return doubleBcdToDecimal(data[3]);
+		}
+
+		inline uint8_t
+		getMonth() {
+			return doubleBcdToDecimal(data[4]);
+		}
+
+		inline uint8_t
+		getDayOfWeek() {
+			return data[5];
+		}
+
+		inline uint8_t
+		getYear() {
+			return data[6];
+		}
+
+		inline bool
+		isWriteProtected() {
+			return data[7] & (1 << 7);
+		}
+
+		///
+		uint8_t data[8];
+
+private:
+		inline uint8_t
+		doubleBcdToDecimal(uint8_t val)
+		{
+			uint8_t tens = ((val & 0x70) >> 4) * 10;
+			uint8_t ones = ( val & 0x0f);
+			return (tens + ones);
+		}
+	};
+};
+
+/**
+ * Realtime clock DS1302
+ *
+ * @ingroup  ds1302
+ * @author   Sascha Schade (strongly-typed)
+ */
+template < class PinSet >
+class Ds1302
+{
+	using Ce   = typename PinSet::Ce;
+	using Sclk = typename PinSet::Sclk;
+	using Io   = typename PinSet::Io;
```

Comment:
This is a good idea, but I think an inherited base struct would be clearer from a documentation view (see https://github.com/roboterclubaachen/xpcc/pull/86#issuecomment-186578136 for demo).  Also the class should probably be called `Config` for conformance with future drivers using this for more than just pins.